### PR TITLE
Fix write_texture for array textures

### LINKED ISSUE #1990: `Queue::write_texture` appears to require row padding for `Texture`s with depth > 1 on Linux/Vulkan
**Description**

Partly to workaround #1672 and partly to achieve better efficiency, I rewrote nannou's `Texture` constructor short-hand functions to use `Queue::write_texture`, rather than creating and submitting a command buffer each time. One of these updated functions loads a 3D texture from a sequence of image buffers with equal dimensions (see [here](https://github.com/nannou-org/nannou/blob/4f797869fca282f3dac756227c091c768a8e7886/nannou_wgpu/src/texture/image.rs#L590-L664)).

After changing this function to use `Queue::write_texture`, I noticed that our `wgpu_image_sequence.rs` example began to show strange behaviour. The example loads a sequence of small PNGs into a single texture to demonstrate how a single 3D texture can be used for animation. When creating an texture view from any layer other than the first, the view appears vertically offset, which suggested to me that the texture's image data layout might be incorrect. Here's example of how the offset looks:

Screenshot from 2021-09-22 10-52-00

Curiously, this only occurs on Linux + Vulkan and doesn't seem to occur on the default macOS and Windows backends.

I noticed that if I manually pad the rows to meet the multiple-of-256-bytes requirement mentioned [in the docs of the `bytes_per_row` field](https://docs.rs/wgpu/0.10.1/wgpu/struct.ImageDataLayout.html#structfield.bytes_per_row), the problem goes away, however the same docs also mention that the requirement should not apply when using `Queue::write_texture`. I can confirm that the requirement doesn't apply when creating single layer textures (these appear to work fine across platforms) - it only appears to be Linux/Vulkan that's the issue.

**Repro steps**

Fetch the branch of [this PR](https://github.com/nannou-org/nannou/pull/795) and `cargo run --example wgpu_image_sequence` from the workspace root on Linux+Vulkan (haven't yet tried Windows+Vulkan).

**Platform**

Linux, GNOME, Wayland, Intel iGPU, Vulkan, wpgu 0.10.1 ([this PR](https://github.com/nannou-org/nannou/pull/795)).

### PR DESCRIPTION
**Pull request by kvark**
**Connections**
Fixes #1990

**Description**
The logic used the original bytes-per-image for offseting the wgpu-hal copy, while it needed to use the staging one.

**Testing**
Not really

### CODE REVIEW

File: wgpu-core/src/device/queue.rs [hunk 1]
```diff
@@ -444,6 +446,10 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
             .map_err(DeviceError::from)?;
         unsafe {
             profiling::scope!("copy");
+            println!(
+                "Copy stage bpr = {}, original = {}",
+                stage_bytes_per_row, bytes_per_row
+            );
```
Comment (cwfitzgerald) [inline wgpu-core/src/device/queue.rs:452]: Errant print

File: wgpu-core/src/device/queue.rs [hunk 2]
```diff
@@ -371,7 +371,9 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         let (selector, dst_base, texture_format) =
             extract_texture_selector(destination, size, &*texture_guard)?;
         let format_desc = texture_format.describe();
-        let (_, bytes_per_array_layer) = validate_linear_texture_data(
+        //Note: `_source_bytes_per_array_layer` is ignored since we have a staging copy,
+        // and it can have a different value.
+        let (_, _source_bytes_per_array_layer) = validate_linear_texture_data(
```
Comment (cwfitzgerald) [inline wgpu-core/src/device/queue.rs:376]: Why aren't we just wholesale ignoring this value (we can leave the comment).

**Review by cwfitzgerald — APPROVED** [review]
Good catch. I tried to take a look at this but couldn't catch the issue.

Nits but algo fine.

File: wgpu-core/src/device/queue.rs (hunk 2 above)
Comment (kvark) [reply wgpu-core/src/device/queue.rs:376]: We could do this. I felt that referring to a field in the comment makes it more clear if the field exists in the code.

File: wgpu-core/src/device/queue.rs (hunk 1 above)
Comment (kvark) [reply wgpu-core/src/device/queue.rs:452]: ooops

**Pull request merged**